tests/stub/iteration/test_iteration_session_run.py

```diff
@@ -61,7 +61,40 @@ def test_error(self):
     def test_all(self):
         self._run(-1, "pull_all.script", ["1", "2", "3", "4", "5", "6"])
 
+    def test_all_slow_connection(self):
+        self._run(-1, "pull_all_slow_connection.script",
+                  ["1", "2", "3", "4", "5", "6"])
+
     # Support -1, not batched at all for BOLTv3
     def test_all_v3(self):
         self._run(-1, "pull_all.script", ["1", "2", "3", "4", "5", "6"],
                   protocol_version="v3")
+
+    def test_discards_on_session_close(self):
+        # TODO remove this block once all languages work
+        if get_driver_name() in ['java']:
+            self.skipTest("Eagerly pulls more results")
```

Comment:
I had a quick look at java.
The reason this test may fail is because it implements a concept of `low and high watermarks`.
In Bolt 4+ the watermarks are defined as follows:
```
highRecordWatermark = (long) (fetchSize * 0.7); // would be 1 when fetchSize is 2
lowRecordWatermark = (long) (fetchSize * 0.3);  // would be 0 when fetchSize is 2
```

The `lowRecordWatermark` is checked when record is returned to the client code:
```
if ( records.size() <= lowRecordWatermark ) {
    .........
    request( fetchSize ); // while in streaming state, this actually is used to calculate how many records to fetch when Success arrives
    .........
}
```

Given the test code, the following happens:
1. Query is submitted to the server
2. First record arrives
3. Because the client is waiting for the first record, its dequeuing begins
4. Low watermark is checked against the internal buffer, which is empty at that point
5. Future fetch is noted internally
6. The first record is returned to the client (client gets unblocked)
7. The internals process the next server responses (RECORD AND SUCCESS)

From this point, depending on timing, at least the following 2 things may happen:
1. If SUCCESS arrives before the client closes the session, the underlying logic will submit another PULL due to the low watermark signal. To account for this, the stub server script may need to support an additional optional PULL.
2. If SUCCESS arrives after the client closes the session, the DISCARD will be sent. This can be reproduced by putting sleep between first and second records in the stub server script.

If this is just one test, we may want to support the additional optional PULL. Otherwise, we may actually want to bring the `watermark feature` awareness to testkit.

Reply:
Solved it by allowing more `PULL`s in the script and asserting only one `PULL` happens for all but the JAVA driver. A long term solution should be found (see Trello).